types/plugin.d.ts

```diff
@@ -11,34 +8,29 @@ export type FastifyPluginOptions = Record<string, any>
  * Fastify allows the user to extend its functionalities with plugins. A plugin can be a set of routes, a server decorator or whatever. To activate plugins, use the `fastify.register()` method.
  */
 export type FastifyPluginCallback<
-  Options extends FastifyPluginOptions = Record<never, never>,
-  Server extends RawServerBase = RawServerDefault,
-  TypeProvider extends FastifyTypeProvider = FastifyTypeProviderDefault,
-  Logger extends FastifyBaseLogger = FastifyBaseLogger
-> = (
-  instance: FastifyInstance<Server, RawRequestDefaultExpression<Server>, RawReplyDefaultExpression<Server>, Logger, TypeProvider>,
-  opts: Options,
-  done: (err?: Error) => void
-) => void
+  Options extends FastifyPluginOptions = FastifyPluginOptions,
+  TIn extends AnyFastifyInstance = AnyFastifyInstance,
+  TOut extends void | AnyFastifyInstance = void | AnyFastifyInstance
+> = (instance: TIn, opts: Options, done: (err?: Error) => void) => TOut
```

Comment:
Are you sure it's ok to break things this way? Don't you think it could surprise users?

Reply:
The question is difficult to answer. What do you mean by "break things this way"?

Reply:
Hi folks! 👋🏼😊 I agree with @jean-michelet, and we should be super careful with such an update, since it looks like a breaking change to me. Considering existing code like:
```ts
  import type {
    FastifyBaseLogger,
    FastifyInstance,
    RawReplyDefaultExpression,
    RawRequestDefaultExpression,
    RawServerDefault,
  } from 'fastify';

  import type { JsonSchemaToTsProvider } from '@fastify/type-provider-json-schema-to-ts';

  const dateSchema = {
    $id: 'date',
    type: 'string',
    format: 'date',
  } as const;

  const jsonReferences = [
    ...dateSchema,
    // all of my JSON Open API schema references ...
  ]

  export type JsonTypeProvider = JsonSchemaToTsProvider<{ references: typeof [] }>;

  export type FastifyJsonSchemaInstance = FastifyInstance<
    RawServerDefault,
    RawRequestDefaultExpression<RawServerDefault>,
    RawReplyDefaultExpression<RawServerDefault>,
    FastifyBaseLogger,
    JsonTypeProvider
  >;
```

This will break the TS build 🤷🏼

Reply:
You can refer to the code snippet above as a test for the change. By applying the updates in this PR, the TS build will fail. With the current `fastify` version, it build passes 😉

Reply:
Alright, i'll set up a reproduction. What error are you seeing, for reference?

Reply:
It's complicated to reproduce it here, I replicated it in a large codebase.
In general, the `FastifyJsonSchemaInstance`, that converts a JSON schema to types, no longer works, so the entire project (that relies on it), has the wrong types when accessing the JSON schemas

Reply:
You can basically check and compare what is returned (as type) by `FastifyJsonSchemaInstance` with this change, and what is returned with the current `fastify` version 😉 You'll see they differ drastically

Reply:
Applying the new version of fastify to fastify-type-provider-json-schema-to-ts does cause type tests to fail in that project, due to the difference in FastifyPluginCallback type params. The type declaration

```
export type FastifyPluginCallbackJsonSchemaToTs<
  Options extends FastifyPluginOptions = Record<never, never>,
  Server extends RawServerBase = RawServerDefault,
  Logger extends FastifyBaseLogger = FastifyBaseLogger,
> = FastifyPluginCallback<Options, Server, JsonSchemaToTsProvider, Logger>;
```

was in general part of the scope that was supposed to be changed since we wanted to remove module augmentation. I will take a look if an interim solution is possible where FastifyPluginCallback can be preserved for this type of scenario.